stablehlo/reference/Element.cpp

```diff
@@ -184,6 +184,30 @@ Element mapWithUpcastToDouble(const Element &el, FloatFn floatFn,
                                      debugString(type).c_str()));
 }
 
+bool compareEqual(APFloat A, APFloat B) {
+  // APFloat::compare treats -0.0 == 0.0, hence following is an additional check
+  // to ensure -0.0 < +0.0.
+  if (A.isZero() && B.isZero() && (A.isNegative() != B.isNegative()))
+    return false;
```

Comment:
Have we established that this is what we need? 

From what I can see, here's how the JAX/XLA pipeline works:
  * JAX uses `lax.eq`, `lax.lt` and `lax.gt` for minmax on complex numbers ([link](https://github.com/google/jax/blob/7c85ca38f45e3ae0d8a80b2d33cf908de9c6564f/jax/_src/lax/lax.py#L2230-L2238)).
  *  These primitives lower directly to `mhlo.compare` with `compare_type = FLOAT` ([link](https://github.com/google/jax/blob/0919a6776afe83400957f2f91c62fd3337c53839/jax/_src/lax/lax.py#L2281-L2303)).
  * In XLA, `kCompare` lowers these further to `FCMP_OEQ`, `FCMP_OLT` and `FCMP_OGT` ([link](https://github.com/tensorflow/tensorflow/blob/29f709d27068b412cbeeecae6507c57462eef64f/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L896-L922)).

Now the question is how the aforementioned LLVM instructions compare `-0.0` and `+0.0`. Does this match `APFloat::compare` (in which case this special case is needed) or not (in which this special case is not needed)?

Methodologically, this is just one implementation of the potential StableHLO spec, so it's not like we must follow it. But we added min/max on complex numbers because JAX supports it, so we might as well completely follow what it does in the draft spec.

Reply:
As per the HLO interpreter,
HLO::kcompare(-0.0, 0.0) == true and 
HLO:kmaximum(-0.0, 0.0) == 0.0

As Jax is using mhlo.maximum which in turn is using kMaximum, we are good with our implementation of max on floating-point type.

For the complex type, as you pointed out, Jax is using kCompare, which as per the HLO interpreter treats 0.0 == -0.0. 

Looking it from the angel of how kCompare is lowered to fcmp, from the following links we can say that fcmp has the same semantics as demonstrated by the interpreter w.r.t.  treating -0.0 == 0.0

https://godbolt.org/z/54c37fTzj : How llc treats the fcmp oeq
https://alive2.llvm.org/ce/z/dzbFCV : How ALive treats `fcmp oeq`. I guess Alive is a well trusted tool for reasoning about LLVM semantics.

Having said that we can drop the special check (to ensure +0.0 > -0.0) and simplify the code,

Reply:
Given that APFloat::compare is unordered and the equality HLO's comparison operation used by HLO in this case is ordered, I never thought if that difference would effect the implementation we have in this PR. Here are my notes:

For an example:
```
func.func @main() -> tensor<complex<f32>> {
  %0 = mhlo.constant dense<(0x7fc0000, 2.5)> : tensor<complex<f32>>
  %1 = mhlo.constant dense<(7.5, 1.5)> : tensor<complex<f32>>
  %2 = mhlo.maximum %0, %1 : tensor<complex<f32>>
  func.return %2 : tensor<complex<f32>>
}
```

Our implementation
```
  auto cmpRes = lhs.real().compare(rhs.real()) == APFloat::cmpEqual
                          ? lhs.imag() > rhs.imag()
                          : lhs.real() > rhs.real();
return cmpRes ? lhs : rhs;
```
`lhs.real().compare(rhs.real())` would return `UNORDERED` and hence `lhs.real().compare(rhs.real()) == APFloat::cmpEqual` is false and hence the `cmpRes` is `lhs.real() > rhs.real()`. Now `ApFloat::operator>` is implemented as 
```
bool operator>(const APFloat &RHS) const {
    return compare(RHS) == cmpGreaterThan;
  }
```
which will return `false` and hence the final return is  `rhs` (`(7.5, 1.5)`).
 
Based on the fact that 
For HLO, for equality comparison they use LLVM's fcmp oeq (see [elemental_ir_emitter.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L904-L905)), the hlo implementation seems to be like

```
auto cmpRes = FCMP_OEQ(lhs.real(),rhs.real())
                          ? FMP_OGTl(hs.imag(), rhs.imag())
                          : FMO_OGT(lhs.real(), rhs.real());
return cmpRes ? lhs : rhs;
```
For the running example,
FCMP_OEQ(lhs.real(),rhs.real())  = `false`
FMO_OGT(lhs.real(), rhs.real()) = `fasle`

leading to the same behavior as StableHLO implementation.